### PR TITLE
Add CBDT support

### PR DESCRIPTION
**Pull request by RoelN**
This PR will add support for CBDT fonts. It will produce a `diff` value in the same way as for vector fonts.

When using HTML output, an animated before/after GIF will be added per modified glyph:

image

The diff routine is optimised in speed and LOC, as comparing CBDT glyphs took a lot of time. This is the only change that also touches vector font diffs, but it produces the exact same diff value.

### CODE REVIEW

**Comment by RoelN** [comment]
Working on tests now!

**Comment by davelab6** [comment]
Overall LGTM, @m4rc1e PTAL

**Comment by guidotheelen** [comment]
@davelab6 @m4rc1e @rsheeter We added some CBDT tests.

**Comment by m4rc1e** [comment]
Thanks for the PR and for tidying up existing parts.

I'm curious, does FreeType not support cbdt tables? it would be better to remove the "image" column from the html table and just generate a separate gif instead

**Comment by anthrotype** [comment]
> does FreeType not support cbdt tables?

yes, it does. And fontdiffenator already depends on freetype-py and Pillow. Here's an example for rendering emoji fonts as PNG with freetype-py + Pillow:

https://github.com/rougier/freetype-py/blob/master/examples/emoji-color.py

File: Lib/diffenator/font.py [hunk 1]
```diff
@@ -56,6 +57,18 @@ class DFont(TTFont):
     def __init__(self, path=None, lazy=False, size=1500):
         self.path = path
         self.ttfont = TTFont(self.path)
+
+        has_outlines = self.ttfont.has_key("glyf") or self.ttfont.has_key("CFF ")
+        if not has_outlines:
+            # Create faux empty glyf table with empty glyphs to make
+            # it a valid font, e.g. for old-style CBDT/CBLC fonts
```
Comment (anthrotype) [inline Lib/diffenator/font.py:64]: @RoelN why is this needed? I think CBDT/CBLC work without glyf table. Can't find reference in OT spec about glyf being required.

Comment (m4rc1e) [reply Lib/diffenator/font.py:64]: He may have added it since Diffenator assumes all fonts need a glyf/cff table (not too sure this is true but seems to make sense).

I'm going through the PR so should have some suggestions regarding this.

Comment (RoelN) [reply Lib/diffenator/font.py:64]: This was indeed added to patch a bitmap font without glyf/cff tables, but only during runnning the script. It will not save the font, but will make the original diffenator code not choke on a glyphless font.

**Comment by RoelN** [comment]
@m4rc1e @anthrotype I was unable to get [set_char_size()](https://github.com/googlefonts/fontdiffenator/pull/74/files#diff-8d56a00e4248b3e3a2cf1399b5a53338R86) to work for Noto Emoji, the CBDT font I'm testing with. Tried just about any size/strike related value. This is why we opted to both patch the font temporarily with a `glyf` table with empty glyphs, and skip some checks when the font is deemed not resizable, which sets apart glyph fonts from CBDT fonts.

Any advice on how to set the pixel size properly is much appreciated!

**Comment by anthrotype** [comment]
> I was unable to get set_char_size() to work for Noto Emoji.

Funnily enough if you use 109 it works.. see @HinTak's comment in
https://github.com/rougier/freetype-py/blob/c578643b5736130c4341584655cb9813c32b5482/examples/emoji-color-cairo.py#L21-L23

probably freetype-py's `set_char_size` should be amended to lookup the closest bitmap strike like `FTDemo_Set_Current_Charsize` in freetype2-demos/src/ftcommon.c is doing.

**Comment by m4rc1e** [comment]
In an ideal world, we'd render fonts containing CBDT tables using FreeType, like we already do for fonts which have a cff or glyf tables. By using this approach, we won't need a separate 'cbdt' section in our reports since they'll just be treated by diffenator as normal glyphs.

@anthrotype has kindly raised an issue in freetype-py, https://github.com/rougier/freetype-py/issues/123 to support CBDT fonts.

If enabling CBDT fonts in freetype is going to be a lot of work, I'm happy to merge this PR and we can do it properly later.

**Comment by RoelN** [comment]
Thanks! I wondered because I tried passing 109 as pixel size and I kept getting the error.

**Comment by anthrotype** [comment]
> Thanks! I wondered because I tried passing 109 as pixel size and I kept getting the error.

it's 109 * 64 since it's a fixed point number

**Comment by HinTak** [comment]
If you run "ftdump " on the font, in the middle of the output is the list of sizes of embedded bitmaps. 160 for Apple and 109 for noto are just conveniently large for visual demo.

In addition, fontVal traps the set_char_size error and output a suitable message too

File: Lib/diffenator/font.py (hunk 1 above)
Comment (davelab6) [reply Lib/diffenator/font.py:64]: > Diffenator assumes all fonts need a glyf/cff table (not too sure this is true but seems to make sense)

It is not true; OpenType fonts must have a glyph table, but that can be glyf, cff, cbdt, sbix, svg

**Comment by RoelN** [comment]
I should explain that we chose for one single before/after GIF per CBDT glyph is to retain as many colors as possible. Putting 2 × 50 full color emoji glyphs inside a before/after GIF with a maximum of 256 colors would make the diff useless.

Even the before/after of a single CBDT glyph reduced to 256 will show color artifacts, let alone if you put 100 of 'em in a GIF.

So that might be an argument to keep the current approach.

**Comment by m4rc1e** [comment]
Good point regarding gif limitations. We could use the apng format instead for diffs since it's 24 bit. I understand this format had patchy support a few years ago but it's apparently much better now.

**Comment by m4rc1e** [comment]
Let's stick with your current approach since this FT setup is going to be a pita. If we decide to support the other color formats, we'll revisit.

I'll take a closer look at the details tomorrow. Thanks.

**Comment by HinTak** [comment]
Fwiw, around the time I added those comments in freetype-py and those lines in FontVal, there was a discussion with Werner about freetype doing the scaling of bitmaps to let set_char_size() work on all sizes. Werner's against that, it should be pango/cairo/pillow doing the scaling instead. I think Behdad was somewhere in the exchange too, and it is probably in the freetype-devel archive.

In any case, the returned error is quite specific - "FT_Error_Invalid_pixel_size" or something liking similar, so you can distinguish fonts having only fixed size bitmaps from other scaling fonts, and other source of error from set_char_size().

So the recommendation from Werner was just to do appropriate things if "invalid_pixel_size" is returned as an error - look for a nearby valid size, or just skip over with an appropriate message (like Fontval does).

Btw, I think apple's emoji contains only two sizes, 40x40 and 160x160 . Google Noto Emoji has a few, like 4-5. Think the mozilla people has a scalable SVG emoji font.

File: Lib/diffenator/__init__.py [hunk 2]
```diff
@@ -27,7 +25,8 @@
     'attribs',
     'metrics',
     'glyphs',
-    'kerns'
+    'kerns',
+    'cbdt'
```
Comment (m4rc1e) [inline Lib/diffenator/__init__.py:29]: "cbdt" can be removed since it isn't a diffenator table.

File: Lib/diffenator/diff.py [hunk 3]
```diff
@@ -17,11 +17,11 @@
 from __future__ import print_function
 import collections
 from diffenator import DiffTable, TXTFormatter, MDFormatter, HTMLFormatter
-import functools
 import os
 import time
 import logging
 from PIL import Image
+from . import read_cbdt
```
Comment (m4rc1e) [inline Lib/diffenator/diff.py:24]: Importing `read_cbdt` can be included in the line 19 import.

File: Lib/diffenator/diff.py [hunk 4]
```diff
@@ -414,30 +436,18 @@ def _diff_images(img_before, img_after):
...
-    data_after = img_after.getdata()
-
-    width, height = max(width_before, width_after), max(height_before, height_after)
-    offset_ax = (width - width_before) // 2
-    offset_ay = (height - height_before) // 2
-    offset_bx = (width - width_after) // 2
-    offset_by = (height - height_after) // 2
+    pixels1 = img_before.convert('L').getdata()
+    pixels2 = img_after.convert('L').getdata()
+    pixels = zip(pixels1, pixels2)
```
Comment (m4rc1e) [inline Lib/diffenator/diff.py:441]: This won't work for images which have different dimensions. Zip's iterator stops when the shortest input iterable is exhausted.

```
>>> a = [1,2,3]
>>> b = [1,2,3,4,5]
>>> zip(a, b)
[(1, 1), (2, 2), (3, 3)]
```

Comment (anthrotype) [reply Lib/diffenator/diff.py:441]: there's https://docs.python.org/3/library/itertools.html#itertools.zip_longest

Comment (m4rc1e) [reply Lib/diffenator/diff.py:441]: That's a great suggestion but it still don't work in this case.

`PIL.Image.convert("L").getdata()` outputs an image's pixels as a single list, `[255, 0, 255...]`. Simply filling the missing values from the smallest list doesn't take into account the image dimensions which is crucial for selecting the correct pixel. In the old code, we use two for loops to get the correct x,y location, we then use these to get the correct index in the list.

**Comment by davelab6** [comment]
@m4rc1e sounds good, please merge and file an issue explaining what needs to be done 'properly' later :)

File: Lib/diffenator/diff.py (hunk 4 above)
Comment (RoelN) [reply Lib/diffenator/diff.py]: I restored the original image diff code!

**Comment by RoelN** [comment]
Remaining issues have been addressed!

**Comment by rsheeter** [comment]
Anything outstanding or can we merge?

**Comment by RoelN** [comment]
Outstanding: using APNG to avoid color loss when putting PNG glyphs in an animated GIF, and getting FreeType  to render CBDT.

I propose making new issues for that, and merging this, as it works and allows you to diff CBDT fonts.

**Comment by m4rc1e** [comment]
I'm going to merge this and push a new release. We can make this better another day.

**Comment by m4rc1e** [comment]
New version pushed to pypi.
